evdns.c

```diff
@@ -1738,18 +1959,173 @@ evdns_add_server_port_with_base(struct event_base *base, evutil_socket_t socket,
-	if (event_add(&port->event, NULL) < 0) {
-		mm_free(port);
-		return NULL;
+	port->max_client_connections = MAX_CLIENT_CONNECTIONS;
+	port->client_connections_count = 0;
+
+	if (sock_type == SOCK_DGRAM) {
+		/* UDP DNS server */
+		event_assign(&port->event, port->event_base,
+					 port->socket, EV_READ | EV_PERSIST,
+					 server_port_ready_callback, port);
+		if (event_add(&port->event, NULL) < 0) {
+			mm_free(port);
+			return NULL;
+		}
+	} else {
+		/* TCP DNS server */
+		port->socket = -1;
+		port->listener = evconnlistener_new(base, incoming_conn_cb, port,
+							LEV_OPT_CLOSE_ON_FREE | LEV_OPT_REUSEABLE, -1, socket);
+		if (!port->listener) {
+			mm_free(port);
+			return NULL;
+		}
 	}
+	LIST_INIT(&port->client_connections);
+
 	EVTHREAD_ALLOC_LOCK(port->lock, EVTHREAD_LOCKTYPE_RECURSIVE);
 	return port;
 }
 
+static void
+server_tcp_event_cb(struct bufferevent *bev, short events, void *ctx);
+
+static int
+tcp_read_message(struct tcp_connection *conn, u8 **msg, int *msg_len)
+{
+	struct bufferevent *bev = conn->bev;
+	struct evbuffer *input = bufferevent_get_input(bev);
+	u8 *packet = NULL;
+	int r = 0;
+
+	EVUTIL_ASSERT(conn);
+	EVUTIL_ASSERT(conn->state == TS_CONNECTED);
+
+	/* reading new packet size */
+	if (!conn->awaiting_packet_size) {
+		if (evbuffer_get_length(input) < 2)
+			goto awaiting_next;
+
+		bufferevent_read(bev, (void*)&conn->awaiting_packet_size,
+			sizeof(conn->awaiting_packet_size));
+		conn->awaiting_packet_size = ntohs(conn->awaiting_packet_size);
+		if (conn->awaiting_packet_size <= 0)
+			goto fail;
+	}
+
+	/* reading new packet content */
+	if (evbuffer_get_length(input) < conn->awaiting_packet_size)
+		goto awaiting_next;
+
+	packet = mm_malloc(conn->awaiting_packet_size);
+	if (!packet)
+		goto fail;
+
+	r = (int)bufferevent_read(bev, (void*)packet, conn->awaiting_packet_size);
+	if (r != conn->awaiting_packet_size) {
+		mm_free (packet);
+		packet = NULL;
+		goto fail;
+	}
+
+	*msg = packet;
+	*msg_len = r;
+awaiting_next:
+	return 0;
+fail:
+	return 1;
+}
+
+static void
+server_tcp_read_packet_cb(struct bufferevent *bev, void *ctx)
+{
+	u8 *msg = NULL;
+	int msg_len = 0;
+	struct client_tcp_connection *client = (struct client_tcp_connection *)ctx;
+	struct evdns_server_port *port = client->port;
+	struct tcp_connection *conn = &client->connection;
+	EVUTIL_ASSERT(port && bev);
+	EVDNS_LOCK(port);
+
+	while (1) {
+		if (tcp_read_message(conn, &msg, &msg_len)) {
+			log(EVDNS_LOG_MSG, "Closing client connection %p due to error", bev);
+			evdns_remove_tcp_client(port, client);
+			EVDNS_UNLOCK(port);
+			if (!port->refcnt)
+				server_port_free(port);
+			return;
+		}
+
+		/* Only part of the message was recieved. */
+		if (!msg)
+			break;
+
+		request_parse(msg, msg_len, port, NULL, 0, client);
+		mm_free(msg);
+		msg = NULL;
+		conn->awaiting_packet_size = 0;
+	}
+
+	bufferevent_setwatermark(bev, EV_READ,
+			conn->awaiting_packet_size ? conn->awaiting_packet_size : 2, 0);
```

Comment:
this file has enough magic numbers, let's use `sizeof(ev_uint16_t)`

Reply:
Done.